### PR TITLE
[feature-wip](recover) new recover ddl and support show dropped

### LINKED ISSUE #8421: [Enhancement] Recover DDL
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

It's serious that overwrite the previous dropped object if drop same object again.
_Caution_: **when create same object after dropped, if wanna recover it, you should rename first**

对于多次删除相同对象会导致垃圾桶中的对象被覆盖(只能恢复最近一次删除)，个人感觉还是蛮危险的事情。
_注意_：**对于删除后又新建同名对象，此时想要恢复之前的对象，请先rename！**

### Solution

* dropped objects with timestamp in trash, it can be query using like `show dropped`
    * fields: id, type(db or table or partition), owner, name, drop_time
* recover ddl: `recover <type> <name> [ using id ] [ to newName ] `
    * can recover to a new name 

* 删除对象可以带时间戳保留在垃圾箱中，并可以通过类似`show dropped`命令查看
    * 大致字段：id, type(db or table or partition), owner, name, drop_time
* 恢复语句可考虑增强：`recover <type> <name> [ using id ] [ to newName ] `
    * 支持恢复到新的名称

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by nextdreamblue**
# Proposed changes

Issue Number: close #8421

## Problem summary

1. 当drop库、表、partition的时候，会放到回收站中，并且允许同名信息，通过库、表、partition的id来区分；
2. 增加show dropped命令，可以查看回收站中库、表、partition的信息，包括类型（db、table、partition）、名称、各级id（比如partition除了显示自身的id，还会显示所属库和表的id），删除时间等；
3. 支持recover的时候指定id去恢复，并且支持恢复到一个新名字；如果不指定id，存在多个同名对象时，会恢复最近一次删除的对象；
4. 另外修复了一个原来bug：当drop 库、表、partition的时候，会记录下删除时间，但是这个删除时间没有第一时间记录到auditlog中，而是需要下一次生成新image的时候才会记录到image中。导致如果中间fe重启，会在回放drop的时候，重新生成一个删除时间，之前这种情况影响不大，无非就是彻底删除时间延迟了；但是现在会根据删除时间去判断恢复哪个表的时候，这个时间就要求非常高，因此此次修改，也优化了这个地方，会在drop相关的auditlog中记录下删除时间。如果是force删除，则记录为0。也因此修改了元数据格式，增加了一个元数据版本114。

后续补充更详细的用户手册或文档、回归测试、测试用例等

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
5. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
6. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
7. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
8. Are there any changes that cannot be rolled back:
    - [x] Yes (If Yes, please explain WHY)
    - [ ] No

修改了与drop相关的journal信息保存的内容，导致不能回滚

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221014174110_clickbench_pr_29429.html

**Comment by morningman** [comment]
I will review this PR

File: fe/fe-core/src/main/cup/sql_parser.cup [hunk 1]
```diff
@@ -3619,6 +3621,10 @@ show_param ::=
     {:
         RESULT = new ShowAnalyzeStmt(tbl, parser.where, orderByClause, limitClause);
     :}
+    | KW_DROPPED opt_wild_where
```
[the remark this answers is not in the document]
Comment (nextdreamblue) [reply fe/fe-core/src/main/cup/sql_parser.cup:3624]: ok, i will fix it

File: fe/fe-core/src/main/java/org/apache/doris/catalog/CatalogRecycleBin.java [hunk 2]
```diff
@@ -76,46 +81,56 @@ public synchronized boolean recycleDatabase(Database db, Set<String> tableNames)
...
+        // erase db with same id
+        eraseDatabaseWithSameId(db.getId());
 
         // recycle db
-        RecycleDatabaseInfo databaseInfo = new RecycleDatabaseInfo(db, tableNames);
+        RecycleDatabaseInfo databaseInfo = new RecycleDatabaseInfo(db, tableNames, tableIds);
         idToDatabase.put(db.getId(), databaseInfo);
-        idToRecycleTime.put(db.getId(), System.currentTimeMillis());
+        if (!isReplay || replayRecycleTime == 0) {
+            recycleTime = System.currentTimeMillis();
```
[the remark this answers is not in the document]
Comment (nextdreamblue) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/CatalogRecycleBin.java:91]: yes, sometimes recycleTime can change when fe restart before saving recycleTime to new image.

File: fe/fe-core/src/main/java/org/apache/doris/analysis/ShowDroppedStmt.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+import org.apache.doris.qe.ShowResultSetMetaData;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+
+import java.util.function.Predicate;
+
+public class ShowDroppedStmt extends ShowStmt {
+    public static final ImmutableList<String> TITLE_NAMES = new ImmutableList.Builder<String>()
+            .add("Type").add("Name").add("DbId").add("TableId").add("PartitionId").add("DropTime")
```
[the remark this answers is not in the document]
Comment (nextdreamblue) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/ShowDroppedStmt.java:35]: 这个地方主要想让用户可以在一个查询中就能获取当前id的所属父id是谁，方便用户去区分同名咨询。比如说两个不同的database，每个database都删除过表test，通过上边的命令，就能看到两个不同的表test属于哪个database。否则很难通过一个过期的id去知道这个表曾经的信息。

File: fe/fe-core/src/main/java/org/apache/doris/catalog/CatalogRecycleBin.java [hunk 4]
```diff
@@ -76,46 +81,56 @@ public synchronized boolean recycleDatabase(Database db, Set<String> tableNames)
         // db should be empty. all tables are recycled before
         Preconditions.checkState(db.getTables().isEmpty());
 
-        // erase db with same name
-        eraseDatabaseWithSameName(db.getFullName());
+        // erase db with same id
+        eraseDatabaseWithSameId(db.getId());
```
[the remark this answers is not in the document]
Comment (nextdreamblue) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/CatalogRecycleBin.java:85]: my fault, i will fix it

File: fe/fe-core/src/main/java/org/apache/doris/catalog/CatalogRecycleBin.java [hunk 5]
```diff
@@ -76,46 +81,56 @@ public synchronized boolean recycleDatabase(Database db, Set<String> tableNames)
...
 
         // recycle db
-        RecycleDatabaseInfo databaseInfo = new RecycleDatabaseInfo(db, tableNames);
+        RecycleDatabaseInfo databaseInfo = new RecycleDatabaseInfo(db, tableNames, tableIds);
```
[the remark this answers is not in the document]
Comment (nextdreamblue) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/CatalogRecycleBin.java:88]: 嗯，后续所有判断都只通过tableIds来做，我优化一下这块逻辑

File: fe/fe-core/src/main/java/org/apache/doris/catalog/CatalogRecycleBin.java [hunk 6]
```diff
@@ -395,9 +463,27 @@ public synchronized void replayRecoverTable(Database db, long tableId) {
             Table table = tableInfo.getTable();
             table.writeLock();
             try {
-                db.createTable(tableInfo.getTable());
+                if (!Strings.isNullOrEmpty(newTableName)) {
```
[the remark this answers is not in the document]
Comment (nextdreamblue) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/CatalogRecycleBin.java:466]: ok, i fix it

File: fe/fe-core/src/main/cup/sql_parser.cup [hunk 7]
```diff
@@ -2743,17 +2745,17 @@ drop_stmt ::=
 
 // Recover statement
 recover_stmt ::=
-    KW_RECOVER KW_DATABASE ident:dbName
+    KW_RECOVER KW_DATABASE ident:dbName opt_id:dbId opt_table_alias:alias
```
[the remark this answers is not in the document]
Comment (nextdreamblue) [reply fe/fe-core/src/main/cup/sql_parser.cup:2748]: ok, i will fix it

[the remark this answers is not in the document]
Comment (nextdreamblue) [reply fe/fe-core/src/main/cup/sql_parser.cup:2748]: 改了一版，发现这个地方如果不写表名的话，很多语法解析比如判断用户是否对该表名有权限很难做，只能从catalogrecyclebin中通过tableid获取到table后才能知道表名，然后才能回头再判断是否读该表有权限去恢复，这太绕了。不如让用户指定表名和id，成对的去恢复，对用户使用也没太大影响

[the remark this answers is not in the document]
Comment (nextdreamblue) [reply fe/fe-core/src/main/cup/sql_parser.cup:2748]: 这块我建议保留现在的设计，让用户自己去输入name和id。只是通过id的话，改动的代码有点多，而且很绕

**Review by morningman — COMMENTED** [review]
LGTM, but please add some regression test for this feature.

**Comment by nextdreamblue** [comment]
> LGTM, but please add some regression test for this feature.

add regression test done

**Review by morningman — APPROVED** [review]
LGTM

**Comment by morningman** [comment]
Still has problem, do not merge
```
2022-10-27 11:56:46,614 WARN (leaderCheckpointer|78) [EditLog.loadJournal():919] [INCONSISTENT META] replay failed  opCode=10 org.apache.doris.persist.CreateTableInfo@55a89c38: errCode = 7, detailMessage = errCode = 2, detailMessage = unknown db, dbName=default_cluster:db1
org.apache.doris.common.MetaNotFoundException: errCode = 7, detailMessage = errCode = 2, detailMessage = unknown db, dbName=default_cluster:db1
	at org.apache.doris.datasource.InternalCatalog.replayCreateTable(InternalCatalog.java:1244) ~[doris-fe.jar:1.0-SNAPSHOT]
	at org.apache.doris.catalog.Env.replayCreateTable(Env.java:3171) ~[doris-fe.jar:1.0-SNAPSHOT]
	at org.apache.doris.persist.EditLog.loadJournal(EditLog.java:200) ~[doris-fe.jar:1.0-SNAPSHOT]
	at org.apache.doris.catalog.Env.replayJournal(Env.java:2457) ~[doris-fe.jar:1.0-SNAPSHOT]
	at org.apache.doris.master.Checkpoint.doCheckpoint(Checkpoint.java:124) ~[doris-fe.jar:1.0-SNAPSHOT]
	at org.apache.doris.master.Checkpoint.runAfterCatalogReady(Checkpoint.java:76) ~[doris-fe.jar:1.0-SNAPSHOT]
	at org.apache.doris.common.util.MasterDaemon.runOneCycle(MasterDaemon.java:58) ~[doris-fe.jar:1.0-SNAPSHOT]
	at org.apache.doris.common.util.Daemon.run(Daemon.java:116) ~[doris-fe.jar:1.0-SNAPSHOT]
```

**Review by morningman — CHANGES_REQUESTED** [review]
some issue

**Comment by nextdreamblue** [comment]
> Still has problem, do not merge
> 
> ```
> 2022-10-27 11:56:46,614 WARN (leaderCheckpointer|78) [EditLog.loadJournal():919] [INCONSISTENT META] replay failed  opCode=10 org.apache.doris.persist.CreateTableInfo@55a89c38: errCode = 7, detailMessage = errCode = 2, detailMessage = unknown db, dbName=default_cluster:db1
> org.apache.doris.common.MetaNotFoundException: errCode = 7, detailMessage = errCode = 2, detailMessage = unknown db, dbName=default_cluster:db1
> 	at org.apache.doris.datasource.InternalCatalog.replayCreateTable(InternalCatalog.java:1244) ~[doris-fe.jar:1.0-SNAPSHOT]
> 	at org.apache.doris.catalog.Env.replayCreateTable(Env.java:3171) ~[doris-fe.jar:1.0-SNAPSHOT]
> 	at org.apache.doris.persist.EditLog.loadJournal(EditLog.java:200) ~[doris-fe.jar:1.0-SNAPSHOT]
> 	at org.apache.doris.catalog.Env.replayJournal(Env.java:2457) ~[doris-fe.jar:1.0-SNAPSHOT]
> 	at org.apache.doris.master.Checkpoint.doCheckpoint(Checkpoint.java:124) ~[doris-fe.jar:1.0-SNAPSHOT]
> 	at org.apache.doris.master.Checkpoint.runAfterCatalogReady(Checkpoint.java:76) ~[doris-fe.jar:1.0-SNAPSHOT]
> 	at org.apache.doris.common.util.MasterDaemon.runOneCycle(MasterDaemon.java:58) ~[doris-fe.jar:1.0-SNAPSHOT]
> 	at org.apache.doris.common.util.Daemon.run(Daemon.java:116) ~[doris-fe.jar:1.0-SNAPSHOT]
> ```

fix it, please try again

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.32 seconds
 load time: 572 seconds
 storage size: 17154644833 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221027180833_clickbench_pr_34730.html

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.72 seconds
 load time: 558 seconds
 storage size: 17154655341 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028023927_clickbench_pr_35041.html

**Review by morningman — DISMISSED** [review]

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.91 seconds
 load time: 561 seconds
 storage size: 17154644846 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028110500_clickbench_pr_35345.html

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**